functorch/functorch/_src/partitioners.py

```diff
@@ -86,6 +87,15 @@ def _is_tangent(node):
 
 def _extract_fwd_bwd_outputs(joint_module: fx.GraphModule):
     num_fwd_outputs = joint_module._out_spec.children_specs[0].num_leaves
+    if hasattr(joint_module, 'input_mutation_submodule'):
+        # When running with functionalization, any program inputs that get updated get converted into outputs
+        # in the forward graph.
+        # This isn't accounted for in the original pytree spec, so we need to include them here.
+        num_input_mutations = len([
+            n for n in joint_module.input_mutation_submodule.graph.nodes
+            if n.op == 'placeholder' and 'primal' in n.target
+        ])
+        num_fwd_outputs += num_input_mutations
```

Comment:
@Chillee let me know how.. suspicious you think this looks. I can't think of a cleaner way to do this; it seems like the partitioning code uses info from the pytree spec on the original joint graph to carve out the forward/backward, and I don't think we want to update that info when we add extra outputs in the forward graph 🤔

Reply:
So there is a bigger question here, which is that when AOTAutograd is wrapped around a chunk of code that makes mutations (or worse, metadata changes) to some inputs, what exactly are we supposed to output in this case.

It seems to me like there are two cases. The easy case is when the mutated inputs don't support grad. Then you can just do the mutation outside of the custom autograd function and you end up with a variable that doesn't support gradients, but that's what the original program would have done anyway, no problem.

The harder case is when the mutated input *does* support grad (because it's like an inplace or something). Custom autograd function can't really deal with this. I'm not sure if there's an easy way to get it to work. also cc @albanD

Reply:
In custom Function, we allow the input to be modified IIF the input is not a view or the custom Function has a single output.
Similarly, if the custom Function returns a view, we mark it such that it cannot be modified inplace in a differentiable manner.

So we do support some cases and raise errors for all the others.

Reply:
Right. But this level of support would be insufficient for arbitrarily whacky partial graphs from dynamo, where any of those situations could occur.

Reply:
> it seems like the partitioning code uses info from the pytree spec on the original joint graph to carve out the forward/backward

I think we should refactor that logic out a bit to be more explicit :)

Reply:
That is correct. But I don't think it is technically possible to support these! (that's another problem of Macro-custom Function!)

Reply:
> In custom Function, we allow the input to be modified IIF the input is not a view or the custom Function has a single output.

> Right. But this level of support would be insufficient for arbitrarily whacky partial graphs from dynamo, where any of those situations could occur.

So, my original goal when writing this PR was that I saw that `torch.dynamo` currently skips `AOTAutograd` entirely [when it detects input mutations](https://github.com/pytorch/torchdynamo/blob/3f2decb5b91d43931606efc72b34f581be839d9d/torchdynamo/optimizations/training.py#L68), and we want to reduce the number of subgraphs that dynamo is forced to fall back on.

Based on the above discussion, it feels wrong to land this PR exactly as is: if `torch.autograd.Function` fundamentally can't support input mutations some cases (that Alban enumerated above), then it seems like AOTAutograd should also properly error in those cases, and dynamo should correctly know to fall back in those cases.

What do you all think of the following?
(1) Update this PR so that AOTAutograd raises an error in the cases Alban pointed out above (if the input we're mutating is a view, and the forward's return type is mult-output)
(2) Some point later, update dynamo to use the same criteria when deciding whether to run the fallback and skip AOTAUtograd

Reply:
cc @anijain2305 @jansel but I think erroring in the backend is too late, assuming `raise_on_backend_error` is on False (which it's not by default), it will still nuke optimizations for the entire graph.

It seems to me like we should AOTAutograd the functional component only, and then strap on a bunch of (unfused) mutations at the tail end that don't get wrapped in AOTAutograd.

Reply:
Coming back to this - using AOTAutograd only on the functional components, and moving the input mutations elsewhere, fully outside of AOTAutograd, makes sense to me.

Models that mutate metadata actually seem like they might not be that uncommon: For example, quantized models like `resnet50_quantized_qat` have running_min/running_max state that get mutated and resized (the ATen op that does this mutating is `_fused_moving_avg_obs_fq_helper`). And the same is probably true for batch norm.

I think my biggest concern is around the API, and how to make it in a way that is (potentially) not BC-breaking for AOTAutograd users.

Some design questions:
(1) Should it be on the callee (dynamo here) to manually call the extra mutation logic? Or should AOTAutograd still arrange for that to happen automatically (I think we probably prefer for this all to still happen transparently, given the complicated calling convention, but let me know if someone thinks differently).
(2) How acceptable is it to change the return type of `aot_function()`?



Brainstorming - the easiest way would probably be to have AOTAutograd always return an `nn.module`, that transparently calls the compiled function (which is an `autograd.Function` object that can't handle certain input mutations, and then manually performs the metadata mutations. E.g. something like this:
```
    class AOTWrapper(nn.Module):
            ...
        def forward(self, *args, **kwargs):
            # compiled_f is the original autograd.Function object that can't handle certain input mutations
            out = compiled_f(*args, **kwargs)
            self.mutate_inputs_submodule(original_inputs, updated_inputs)
```